Rework management of Salt jobs in the UI

gdemonet opened the pull request:
**Component**: ui

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**:

As the UI will need to handle more and more long-running Salt operations (such as importing a Solution or preparing an Environment), the pattern introduced for deployment of Nodes needs to be extended.

**Summary**:

- extract all logic related to management of Salt jobs from `app.nodes` to `app.salt`
- refine the pattern to make it more generic: a deployment job is now named `deploy-node/<node-name>` to avoid future collisions
- store the completion time of a job in localStorage, to provide efficient reloading of the application
- keep jobs in localStorage after their completion and implement a garbage collection saga to clean them up after some time
- trigger notifications from the related duck (here, `app.nodes`), while the action gets sent from the `app.salt` duck
- rework the NodeDeployment page slightly (simplification and handling of some edge-cases)

(read commit messages for more info)

**Acceptance criteria**: 

Everything still works, and one can now find info about previously failed jobs even after a refresh. Unit tests were also slightly reworked to make use of features of `jest` to avoid code duplication - make sure you (UI devs) understand how this works and feel comfortable with it.

gdemonet committed:
ui/ducks/nodes: Apply Prettier

Run Prettier on this file using the committed `ui/.prettierrc`, that
every developer should use for consistency.

gdemonet committed:
ui/ducks/nodes: Spawn all sagas in parallel

Even though yielding a `takeEvery` effect is non-blocking, hence doesn't
create any issues, it's not the intention of the code. One actually
wants to spawn `all` sagas in parallel.

gdemonet committed:
ui/ducks/salt: Rework management of Jobs

The initial implementation was focused on Jobs for the deployment of new
Nodes, but this was both limiting (as we cannot re-use it for other
kinds of Salt Jobs) and confusing (actions and sagas in the `nodes` duck
were not actually about Nodes per-se).

We introduce a more generic solution, that decouples the creation,
persistence, and observation of active Jobs from the consumption of the
Jobs' status and result. The `app.salt` duck is re-purposed for this
use-case.

Note that some interesting saga patterns are introduced to cope with the
complexity of this management (buffering of actions to manage a
persisted state in LocalStorage, garbage collection of old persisted
Jobs...).

Also note that, while adapting the NodeDeployment page to this
refactored state (as it is the main consumer of Jobs status and result),
we drastically simplify it: there was no need for an internal state of
the component, as all the required information is available in the Redux
state.

gdemonet committed:
ui/nodes/deploy: Handle missing Job info

Now that the NodeDeployment page is based on Node name, and not some
JID, it seems reasonable that users will bookmark it, or at least
refresh it.

We thus handle the following edge-cases:
- no existing deployment Job for this Node in the localStorage
- refreshing information about a Job stored in the localStorage

gdemonet committed:
ui/nodes/deploy: Change route path

Instead of `/nodes/deploy/$name`, we go for `/nodes/$name/deploy`, for
consistency with other routes in the same `/nodes` namespace.

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/rework-ui-salt-jobs` into
`development/2.4`
* [w/2.5/improvement/rework-ui-salt-jobs](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/rework-ui-salt-jobs) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

gdemonet committed:
ui/ducks: Define unique selectors

This will help consumers (in case a duck is moved), and testers (the
select() effect can be used and its selector checked).

gdemonet committed:
ui/nodes/test: Refactor tests for clarity

Instead of duplicating large test methods while only changing a few
details about the test input, let's rely on the `test.each` helper
instead.
This helps readers understand what the purpose of a test is, and the
different cases it needs to support.

gdemonet committed:
ui/tests: Make sure JUnit report shows useful info

gdemonet committed:
ui/salt/tests: Refactor utils tests for clarity

gdemonet committed:
ui/tests: Setup localStorage mock

To provide easier manipulation of localStorage in our unit tests, we
setup a mock (which implementation is super naive, and gets inspiration
from `github.com/clarkbw/jest-localstorage-mock`).

gdemonet committed:
ui/salt/tests: Test the newly introduced sagas

The sagas for managing Salt jobs (in `state.app.salt.jobs`) need good
testing coverage, as they will be the backbone for multiple parts of the
application that rely on those jobs.

Note that a number of constants and functions needed to be exported from
the original duck just for their usage in tests, which is why the
exports are done in this commit.

gdemonet committed:
ui/nodes/test: Introduce cloneableGenerator

This helper allows to reduce some code duplication when going through
the first steps of a saga generator. It doesn't completely removes the
need to manually operate on the "last" steps in each test case, which is
factorized here using a custom helper `_checkCompletion(generator)`.

gdemonet committed:
ui/nodes/test: Test how `fetchNodes` reads Jobs

In order to mark Nodes as "deploying" or not, the `fetchNodes` saga
needs to read the list of Jobs in the state and map each active job to
the corresponding Node (based on node name, compared to Job name, the
latter being expected to be of the form "deploy-node/<node-name>").

Some extra test cases are marked as `todo`, but the nominal path is
covered.

bert-e commented:
# History mismatch



Merge commit #365d392879bc7dd11f1ecc9a172a0ece5f6e3dc2 on the integration branch
`w/2.5/improvement/rework-ui-salt-jobs` is merging a branch which is neither the current
branch `improvement/rework-ui-salt-jobs` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/rework-ui-salt-jobs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/rework-ui-salt-jobs` into
`development/2.4`
* [w/2.5/improvement/rework-ui-salt-jobs](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/rework-ui-salt-jobs) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #253b8e6508c76bd4bdaa7517f8d26ac4873c8534 on the integration branch
`w/2.5/improvement/rework-ui-salt-jobs` is merging a branch which is neither the current
branch `improvement/rework-ui-salt-jobs` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/rework-ui-salt-jobs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/rework-ui-salt-jobs` into
`development/2.4`
* [w/2.5/improvement/rework-ui-salt-jobs](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/rework-ui-salt-jobs) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

MonPote commented on ui/src/ducks/app/salt.js:
```diff
@@ -1,31 +1,255 @@
+  // NOTE: instead of creating a saga per action type, since the operations on
+  // localStorage rely on its previous state, we need to execute them serially.
+  // We don't set up a buffer size as operations should be infrequent and fast.
+  const channel = actionChannel([ADD_JOB, REMOVE_JOB, JOB_COMPLETED]);
```
Very cool pattern :)

carlito-scality commented on ui/src/containers/NodeDeployment.js:
```diff
@@ -140,14 +129,26 @@ const NodeDeployment = ({ intl }) => {
-        </NodeDeploymentContent>
+        {activeJob === undefined ? (
+          <InfoMessage>
+            {`No deployment job found for Node ${nodeName}.`}
```
translation here

carlito-scality commented on ui/src/containers/NodeDeployment.js:
```diff
@@ -140,14 +129,26 @@ const NodeDeployment = ({ intl }) => {
+            {`No deployment job found for Node ${nodeName}.`}
+          </InfoMessage>
+        ) : activeJob.completed && isEmpty(activeJob.status) ? (
+          <InfoMessage>{`Refreshing job info...`}</InfoMessage>
```
translation here

carlito-scality commented on ui/src/ducks/app/nodes.js:
```diff
@@ -434,104 +361,60 @@ export function* deployNode({ payload }) {
+    yield put(
+      addJobAction({
+        name: `deploy-node/${payload.name}`,
+        jid: result.return[0].jid,
```
why don't we add `node: payload.name` to avoid `job.name.replace(/^(deploy-node\/)/, '')`?

gdemonet replied:
I actually didn't want to add too much in `localStorage`. But since we don't care too much about "dimensions" when it comes to JSON, I think it actually makes a lot of sense to switch to `type` + some arbitrary number of attributes for consumers to read from, in place of some known format to encode these attributes in a single `name` string.

Good point :)

carlito-scality commented on ui/src/services/salt/utils.js:
```diff
@@ -1,7 +1,50 @@
+  if (job) {
+    job.completedAt = completedAt;
+  }
+  localStorage.setItem(JOBS, JSON.stringify(jobs));
```
to be replaced by `saveJobsToLocalStorage(jobs)`

carlito-scality commented on ui/src/ducks/app/salt.js:
```diff
@@ -1,31 +1,258 @@
+  const result = yield call(ApiSalt.printJob, job.jid);
+  // TODO: error handling?
+  const status = getJobStatusFromPrintJob(result, job.jid);
+  return yield* updateJobStatus(job, status);
```
why `yield*` instead id `yield call`?

gdemonet replied:
:man_shrugging: I read that `yield*` is the idiomatic way of doing this with ES6 generators if you want sequential execution. But true, `call` is useful if we want to test that this saga delegates to another.

carlito-scality commented on ui/src/ducks/app/salt.js:
```diff
@@ -1,31 +1,258 @@
+    fork(manageLocalStorage),
+    fork(garbageCollectJobs),
+    takeEvery(CONNECT_SALT_API, initialize),
+    takeEvery(CONNECT_SALT_API, watchSaltEvents),
```
I would merge those two in one saga and call `initialize` before `watchSaltEvents ` so we are sure we do not miss active jobs

gdemonet replied:
Good point! I'll do that

carlito-scality commented on ui/src/ducks/app/salt.js:
```diff
@@ -1,31 +1,258 @@
+  const payload = { ...defaultJob, ...job };
+  if (payload.completedAt !== null) {
+    payload.completed = true;
+  }
```
I wonder if it is necessary here since we will overwrite it in `updateJobStatus`

gdemonet replied:
It's for the `refreshNodes` saga actually, so that they don't show the node as deploying while we wait for an update of the status. We know the job is completed, and the consumer uses `completed`, not `completedAt` :/

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @carlito-scality

gdemonet committed:
ui/salt: Allow arbitrary props in Job objects

We previously used a single prop, `name`, to encode information about
the Job at hand. For Node deployment Jobs, we used the
`deploy-node/<nodeName>` pattern, which is cumbersome to manipulate.

Since this is JSON, there is no overhead to add optional dimensions, and
there will be close to no impact on localStorage size.

bert-e commented:
# History mismatch



Merge commit #de03f2166dd4ff2dace2c2559f9302f1d4f5c517 on the integration branch
`w/2.5/improvement/rework-ui-salt-jobs` is merging a branch which is neither the current
branch `improvement/rework-ui-salt-jobs` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/rework-ui-salt-jobs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/rework-ui-salt-jobs` into
`development/2.4`
* [w/2.5/improvement/rework-ui-salt-jobs](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/rework-ui-salt-jobs) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @carlito-scality

carlito-scality commented on ui/src/ducks/app/salt.js:
```diff
@@ -1,31 +1,267 @@
+export const allJobsSelector = state => state.app.salt.jobs;
+
+// Sagas
+export function* initialize() {
```
you need to get a `payload` here and pass it to `yield fork(watchSaltEvents, payload);`

bert-e commented:
# History mismatch



Merge commit #0a465c502245888a1ecd01c4b1536919b734764f on the integration branch
`w/2.5/improvement/rework-ui-salt-jobs` is merging a branch which is neither the current
branch `improvement/rework-ui-salt-jobs` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/rework-ui-salt-jobs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/rework-ui-salt-jobs` into
`development/2.4`
* [w/2.5/improvement/rework-ui-salt-jobs](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/rework-ui-salt-jobs) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @carlito-scality

gdemonet committed:
ui/nodes/deploy: Use intl.translate from global provider

bert-e commented:
# History mismatch



Merge commit #62d5fd16cc3edb0c099b52ea8a3019142591672a on the integration branch
`w/2.5/improvement/rework-ui-salt-jobs` is merging a branch which is neither the current
branch `improvement/rework-ui-salt-jobs` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/rework-ui-salt-jobs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/rework-ui-salt-jobs` with contents from `improvement/rework-ui-salt-jobs`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/rework-ui-salt-jobs origin/development/2.5
 $ git merge origin/improvement/rework-ui-salt-jobs
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/rework-ui-salt-jobs
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @carlito-scality

bert-e commented:
# History mismatch



Merge commit #253c17add83a9b2a40c225d475eabf14d2e5c9d6 on the integration branch
`w/2.5/improvement/rework-ui-salt-jobs` is merging a branch which is neither the current
branch `improvement/rework-ui-salt-jobs` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/rework-ui-salt-jobs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @carlito-scality







*The following options are set:* **approve**

carlito-scality reviewed (approved):
LGTM

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/6b717e27394003fa4113f00145c581807602e0a1) did not succeed in branch w/2.5/improvement/rework-ui-salt-jobs.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/ce0c1a894645d5291ecf05e6c68053686358dfc5) did not succeed in branch w/2.5/improvement/rework-ui-salt-jobs.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
